sycl/plugins/cuda/pi_cuda.cpp

```diff
@@ -528,43 +530,52 @@ pi_result cuda_piPlatformsGet(pi_uint32 num_entries, pi_platform *platforms,
                               pi_uint32 *num_platforms) {
 
   try {
-    static constexpr pi_uint32 numPlatforms = 1;
+    static std::once_flag initFlag;
+    static pi_uint32 numPlatforms = 1;
+    static _pi_platform platformId;
 
-    if (num_platforms != nullptr) {
-      *num_platforms = numPlatforms;
+    if (num_entries == 0 and platforms != nullptr) {
+      return PI_INVALID_VALUE;
+    }
+    if (platforms == nullptr and num_platforms == nullptr) {
+      return PI_INVALID_VALUE;
     }
 
     pi_result err = PI_SUCCESS;
 
-    if (platforms != nullptr) {
-
-      assert(num_entries != 0);
-
-      static std::once_flag initFlag;
-      static _pi_platform platformId;
-      std::call_once(
-          initFlag,
-          [](pi_result &err) {
-            err = PI_CHECK_ERROR(cuInit(0));
-
-            int numDevices = 0;
-            err = PI_CHECK_ERROR(cuDeviceGetCount(&numDevices));
-            platformId.devices_.reserve(numDevices);
-            try {
-              for (int i = 0; i < numDevices; ++i) {
-                CUdevice device;
-                err = PI_CHECK_ERROR(cuDeviceGet(&device, i));
-                platformId.devices_.emplace_back(
-                    new _pi_device{device, &platformId});
-              }
-            } catch (...) {
-              // Clear and rethrow to allow retry
-              platformId.devices_.clear();
-              throw;
+    std::call_once(
```

Comment:
In the original code, the `call_once` that triggers `cuInit` is never called if the SYCL RT doesn't pass the `pi_platform *platforms` parameter, e.g. when just querying the number of platforms available.
In this patch,  `cuInit` is always called, even when just querying the number of platforms. 

This maybe breaks some unit tests for the pi cuda backend, but they are disabled at the moment. @bjoernknafla / @Alexander-Johnston  : maybe make a note of this change for when resuming the work on the unit tests?

Reply:
One difference will be that, on a machine without CUDA devices, this will now return 0 platforms.

Reply:
Famous last words though I do not think that calling `cuInit` earlier than before will have an effect on the unit tests (the `call_once` should be enough from stopping bad behavior).

Reply:
We have a not yet shared patch for dealing with not finding platforms and/or devices.

Reply:
PS: I like this change a lot!